[Ng Xiang Jun] iP

xiangjunn opened the pull request:
### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2122S1/ip/pull/164).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😁 

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

xiangjunn committed:
Add greeting

xiangjunn committed:
Add echo feature that exits program when user types bye

xiangjunn committed:
Add the ability to store text entered by the user

xiangjunn committed:
Display tasks to user when requested

xiangjunn committed:
Create a task class to represent tasks

xiangjunn committed:
Improve code logic

xiangjunn committed:
Add the ability to mark tasks as done

xiangjunn committed:
Add support for tracking three types of tasks

xiangjunn committed:
Remove unnecessary print statement in Duke.java

xiangjunn committed:
Implement automated text ui testing

xiangjunn committed:
Handle errors when user enters incorrect input

xiangjunn committed:
Update test case

xiangjunn committed:
Add support for deleting tasks from the list

xiangjunn committed:
Improve logic for adding task

xiangjunn committed:
Update test case to check for delete

xiangjunn committed:
Update code to work with Command enum

xiangjunn committed:
Update EXPECTED.TXT

xiangjunn committed:
Ensure duke.txt exists when Duke starts up

xiangjunn committed:
Save the task to duke.txt when adding task

xiangjunn committed:
Change implementation for storing tasks

xiangjunn committed:
Load data from hard disk when Duke starts up

xiangjunn committed:
Store dates as LocalDate instead of String in deadline and event

xiangjunn committed:
Merge branch 'branch-Level-7'

xiangjunn committed:
Resolve merge conflict by updating codebase affected by changes

xiangjunn committed:
Make code neater

xiangjunn committed:
Merge branch 'branch-Level-8'

xiangjunn committed:
Rename Command enum to CommandKeyword

xiangjunn committed:
Create Command class to manage commands

xiangjunn committed:
Create Ui class to deal with interactions with user

xiangjunn committed:
Implement Storage class to deal with loading/saving tasks in file

xiangjunn committed:
Create Parser class to deal with making sense of the user command

xiangjunn committed:
Create Tasklist class with operations to change the list

xiangjunn committed:
Update Duke.java to work with refactored code

xiangjunn committed:
Divide classes into packages

xiangjunn committed:
Update runtest.bat to work with new package

xiangjunn committed:
Add JUnit tests

xiangjunn committed:
Modify .gitignore to ignore MANIFES.MF

xiangjunn committed:
Add JavaDoc comments

xiangjunn committed:
Tweak the code to comply with coding standard

xiangjunn committed:
Add command to find a task by searching for keyword

xiangjunn committed:
Revert "Add JavaDoc comments"

This reverts commit f9d582edb44081536afc7e67977fae1a1fe8c1e5.

xiangjunn committed:
Update all constructors' documentation

xiangjunn committed:
Merge branch 'branch-A-JavaDoc'

xiangjunn committed:
Merge branch 'branch-A-CodingStandard'

xiangjunn committed:
Merge branch 'branch-Level-9'

xiangjunn committed:
Fix raw type ArrayList to ArrayList<String>

xiangjunn committed:
Update text-ui-test

xiangjunn committed:
Merge remote-tracking branch 'origin/add-gradle-support'

xiangjunn committed:
Add missing JavaDocs

xiangjunn committed:
Edit code to pass checkstyle

xiangjunn committed:
Merge branch 'branch-A-CheckStyle'

xiangjunn committed:
Update test files to adhere to checkstyle

YoYoCiti commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+                file.createNewFile();
+            }
+        } catch (IOException e) {
+            // this exception should not occur because the input(path) is fixed
```
May still be safer to handle this exception since there may be unforeseen issues with the file on another system?

YoYoCiti commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,158 @@
+    public Ui() {
+        this.sc = new Scanner(System.in);
+    }
+    private void printMessage(String message) {
```
```suggestion
    
    private void printMessage(String message) {
```

YoYoCiti commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public String stringToStore() {
+        return Deadline.TASK_LETTER + " | " + this.getStatusIcon() + " | " + this.description + " | "
+                + this.date.format(Deadline.DATE_TIME_FORMATTER) + " | " + this.time + "\n";
```
```suggestion
        return String.format("%s | %s | %s | %s | %s \n", Deadline.TASK_LETTER, this.getStatusIcon(), 
                this.description, this.date.format(Deadline.DATE_TIME_FORMATTER), this.time);
```

Could consider for better readability?

YoYoCiti commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+        tasks.add(task);
+        this.ui.showAddTask(task, this.tasks.size());
+        return this.tasks;
+    }
```
This method is pretty long. Maybe it would be cleaner to catch and throw the exceptions in the different Task classes (i.e Event, Deadline etc.) as the responsibility to do so seems to be more of theirs rather than the TaskList.

YoYoCiti reviewed (commented):
Overall, easy to read and good quality code :smile: Only a few nitpicks. Thanks.

ganhongyao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+public class Parser {
+
+    /**
+     * Execute user command and returns a boolean that represents the continuation of the program.
```
Should the comment start with "Executes" instead of "Execute"? I noticed this in several other places too.
```suggestion
     * Executes user command and returns a boolean that represents the continuation of the program.
```

ganhongyao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+            tl.find(restOfCommand);
+            break;
+        default:
+            /* will never be executed because the error would have been caught in run() method
+               if the user input a command that is invalid */
```
I like how you explicitly state this behaviour which may not be obvious at first glance 👍

ganhongyao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+    public Storage(Ui ui) {
+        this.ui = ui;
+        try {
+            String dir = System.getProperty("user.dir");
+            Path path = Paths.get(dir, "data");
+
+            // create the data folder if it does not exist
+            if (!Files.exists(path)) {
+                path = Files.createDirectory(path);
+            }
+
+            File file = path.resolve("duke.txt").toFile();
```
I like how you separate the different logical units using line breaks. 👍

ganhongyao commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+     * @throws DukeException If the task fails to be added to the list.
+     */
+    public ArrayList<Task> addTask(String description, CommandKeyword keyword) throws DukeException {
+        Task task = null;
```
Would it be better if you do the declaration and initialisation together in the respective case blocks, and doing `tasks.add(task)` in the case block as well? I think that this has the benefit of ensuring that the task will never be `null`, especially if the logic is changed in some way in future.

ganhongyao commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+     * @param ui An Ui instance to deal with interactions with the user.
+     */
+    public TaskList(ArrayList<String> tasksInfo, Ui ui) {
+        List<Task> list = tasksInfo.stream().map(Task::convertToTask).collect(Collectors.toList());
```
Would this be a more descriptive variable name?
```suggestion
        List<Task> tasks = tasksInfo.stream().map(Task::convertToTask).collect(Collectors.toList());
```

ganhongyao reviewed (commented):
Overall, your code looks well-organised and easy to follow! 👍 Added some minor comments.

xiangjunn committed:
Add basic GUI for Duke without functionality

xiangjunn committed:
Include functionality with some improvements in GUI

xiangjunn committed:
Remove sad face emoji from error messages

The emojis are shown up as ugly characters in GUI.

Error messages that contain emojis were updated to not contain them.

Removing them will improve user experience.

xiangjunn committed:
Merge branch 'branch-Level-10'

xiangjunn committed:
Add assert feature

Document important assumptions that should hold at
various points in the code.

An assertion failure indicates the programmer made
a mistake in the code and he can correct it.

Impact of assertions on performance is considered low
and worth the additional safety they provide.

xiangjunn committed:
Improve on code quality

Code quality helps to improve readability, which can
save programmers a lot of time when they are debugging.

Remove dead code, structure code logically and reduce
the size of long methods.

There are a few methods that are very long, such as
parse method in Parser class and addTask in TaskList
class. By splitting the codes to more methods, it is
easier to read and update changes.

xiangjunn committed:
Merge pull request #1 from xiangjunn/branch-A-Assertions

Add assert feature

xiangjunn committed:
Merge branch 'master' into branch-A-CodeQuality

xiangjunn committed:
Merge pull request #2 from xiangjunn/branch-A-CodeQuality

Improve on code quality

xiangjunn committed:
Create gradle.yml

xiangjunn committed:
Add friendlier syntax for commands

xiangjunn committed:
Add JavaDoc for new methods

xiangjunn committed:
Edit code to follow checkstyle

xiangjunn committed:
Merge pull request #3 from xiangjunn/branch-C-FriendlierSyntax

Add friendlier syntax for commands

xiangjunn committed:
Add background color to textbox

xiangjunn committed:
Fix delete command not working bug

xiangjunn committed:
Improve GUI

xiangjunn committed:
Add background color for dialogContainer

xiangjunn committed:
Improve GUI for button and textfield

xiangjunn committed:
Differentiate background color for error and normal messages

xiangjunn committed:
Merge pull request #5 from xiangjunn/branch-A-BetterGUI

Better GUI

xiangjunn committed:
Update JavaDoc comments

xiangjunn committed:
Add a representative screenshot of the product

xiangjunn committed:
Add Quick start to user guide

xiangjunn committed:
Finish up User Guide

xiangjunn committed:
Merge pull request #6 from xiangjunn/branch-A-UserGuide

User Guide

xiangjunn committed:
Update user guide

xiangjunn committed:
Set the screen to be not resizable

xiangjunn committed:
Align text to display picture side

xiangjunn committed:
Divide the files into more folders

xiangjunn committed:
Add Equal method in Command class

xiangjunn committed:
Add more test cases

xiangjunn committed:
Rename app name from duke to SaitamaChat

xiangjunn committed:
Update Ui.png
